src/store/participantsStore.js

```diff
@@ -821,31 +863,91 @@ const actions = {
 			return
 		}
 
-		commit('setInCall', {
-			token,
-			sessionId: participantIdentifier.sessionId,
-			flags,
-		})
+		let isParticipantsListReceived = null
+		commit('joiningCall', { token, sessionId, flags })
 
-		const actualFlags = await joinCall(token, flags, silent, recordingConsent)
+		const handleJoinCall = ([token, flags]) => {
+			commit('setInCall', { token, sessionId, flags })
+			commit('finishedJoiningCall', { token, sessionId })
 
-		const updatedData = {
-			inCall: actualFlags,
+			if (isParticipantsListReceived) {
+				finishConnecting()
+			} else {
+				commit('connecting', { token, sessionId, flags })
+			}
 		}
-		commit('updateParticipant', { token, attendeeId: attendee.attendeeId, updatedData })
 
-		EventBus.once('signaling-users-in-room', () => {
-			commit('finishedConnecting', { token, sessionId: participantIdentifier.sessionId })
-		})
+		const handleJoinCallFailed = ([token, payload]) => {
+			finishConnecting()
+			commit('connectionFailed', {
+				token,
+				payload
+			})
+			commit('setInCall', {
+				token,
+				sessionId: participantIdentifier.sessionId,
+				flags: PARTICIPANT.CALL_FLAG.DISCONNECTED,
+			})
+		}
+
+		const handleUsersInRoom = (payload) => {
+			const participant = payload[0].find(p => p.sessionId === sessionId)
+			if (participant && participant.inCall !== PARTICIPANT.CALL_FLAG.DISCONNECTED) {
+				if (state.joiningCall[token]?.[sessionId]) {
+					isParticipantsListReceived = true
+					commit('connecting', { token, sessionId, flags })
+					return
+				}
+				finishConnecting()
+			}
+		}
+
+		const handleUsersChanged = (payload) => {
+			const participant = payload[0].find(p => p.nextcloudSessionId === sessionId)
+			if (participant && participant.inCall !== PARTICIPANT.CALL_FLAG.DISCONNECTED) {
+				if (state.joiningCall[token]?.[sessionId]) {
+					isParticipantsListReceived = true
+					commit('connecting', { token, sessionId, flags })
+					return
+				}
+				finishConnecting()
+			}
+		}
+
+		const finishConnecting = () => {
+			isParticipantsListReceived = null
```

Comment:
Maybe we don't need to reassign it back to null/false, since it's purpose is completed, and we won't change it later anymore in this function scope
```suggestion
```